From 1ee23bd8c37d5645005906cf93841452a28bae52 Mon Sep 17 00:00:00 2001
From: Laxminath Kasam <lkasam@codeaurora.org>
Date: Thu, 9 Jul 2020 19:36:36 +0530
Subject: [PATCH] asoc: Reset the buffer if size is partial or zero

Sometimes during device switch in recording,
observe size 0 is return from DSP due to EOS
handling. For ALSA pcm_read to unblock, buffer
appl_ptr is elapsed without actually updating
the buffer. And userspace copies the stale
data(old buffer) causing issue sometimes.
Reset the buffer for that period_size in
such cases instead of transfer stale data.

Change-Id: I0d3ac133a8d95fad0710586e3e947410a41c9c5a
Signed-off-by: Laxminath Kasam <lkasam@codeaurora.org>
---
 techpack/audio/asoc/msm-pcm-q6-v2.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/techpack/audio/asoc/msm-pcm-q6-v2.c b/techpack/audio/asoc/msm-pcm-q6-v2.c
index 6f4a0aa213bd..ebcd54854513 100755
--- a/techpack/audio/asoc/msm-pcm-q6-v2.c
+++ b/techpack/audio/asoc/msm-pcm-q6-v2.c
@@ -953,6 +953,11 @@ static int msm_pcm_capture_copy(struct snd_pcm_substream *substream,
 			goto fail;
 		}
 
+		if (size == 0 || size < fbytes) {
+			memset(bufptr + offset + size, 0, fbytes - size);
+			size = xfer = fbytes;
+		}
+
 		if (copy_to_user(buf, bufptr+offset, xfer)) {
 			pr_err("Failed to copy buf to user\n");
 			ret = -EFAULT;
